python/phase-1.md

```diff
@@ -0,0 +1,380 @@
+# Python SDK - Phase 1
+
+The Python SDK will be implemented in three loosely defined phases:
+
+* Phase 1 - Initial project, activity worker, and client
+* Phase 2 - Workflow worker, local activities, async activity completion, etc
+* Phase 3 - Deterministic workflow sandbox, replayer, test suite (some parts may be combined with phase 2)
+
+This proposal is for phase 1. It is intentionally loosely defined and all aspects are subject to discussion/change.
+
+## Overview
+
+The Python SDK will be backed by [sdk-core](https://github.com/temporalio/sdk-core), utilizing a bridge based on
+[PyO3](https://github.com/PyO3/PyO3) similar to `sdk-typescript`'s `neon`-based bridge.
+
+The high-level goals for the Python SDK are:
+
+* Be as simple as possible from a caller POV
+* Be as lightweight as possible wrt dependencies
+* Be reasonably flexible with reasonable defaults
+* Support only Python >= 3.7
+* Be friendly to REPL/interpreter use cases
+
+The stack/approach looks like:
+
+* Python >= 3.7 (so may need `typing_extension` for things like `ParamSpec`)
+* Poetry packaging, black formatting
+  * Pipenv originally considered
+* PyO3 + sdk-core
+* asyncio direct (i.e. no need for anyio-style wrapper)
+* Type annotations everywhere
+* Decorator-based activity and workflow definitions
+
+An example of a client (intentionally missing _a lot_ of features):
+
+```python
+import temporalio
+
+def main():
+  client = temporalio.Client("localhost:7233", namespace="MyNamespace")
```

Comment:
We might need to change this to `temporalio.Client.create` so it can be `await`d on.

Reply:
I don't think we want to expose `Client` here, I'd rather stick to what we have in Java and TS where there's a grpc client (with a small wrapper to create it), a Worker and a WorkflowClient.

Reply:
This is mostly the `WorkflowClient` from the Java and TS perspective, I am just calling it "Client" for clarity because I think that's clearer and lets us add things like async activity heartbeat/completion to it.

I am trying to have the simplest API possible, which means I am trying to avoid two-step client creation. But if y'all really want, I can make a gRPC client, a workflow client, an async activity client, etc. It just feels bloated/unnecessary.

Reply:
I'm fine with unifying the async activity completion client and workflow client into a single `Client` class.

Reply:
Yay! Thanks.

Reply:
Updated